### PR TITLE
util and ephemeral test

### PR DESCRIPTION
**Pull request by levente232**

### CODE REVIEW

File: ansible/playbooks/reset_moriod.yml [hunk 1]
```diff
@@ -11,34 +11,69 @@
...
       ignore_errors: true
 
-    - name: Remove Morio database files
+    - name: Find all database files in the directory
+      ansible.builtin.find:
+        paths: /var/lib/morio/moriod
+        file_type: file
+      register: db_files
+
+    - name: Remove all database files
```
Comment (joostdecock) [inline ansible/playbooks/reset_moriod.yml]: It would be more accurate to write `Remove all data files` because it's more than just the database

File: ansible/playbooks/reset_moriod.yml [hunk 2]
```diff
@@ -11,34 +11,69 @@
...
       ignore_errors: true
 
-    - name: Backup version info
+    - name: Find version files
+      ansible.builtin.find:
+        paths: /etc/morio/moriod
+        patterns: version*
+      register: version_files
+
+    - name: Backup version files
```
Comment (joostdecock) [inline ansible/playbooks/reset_moriod.yml:31]: Is there are specific reason we are splitting this step in two parts (find and copy)?
What I mean is, what is the benefit of splitting this step over handling the copy in one step like this:
```yaml
- name: Backup version files
  ansible.builtin.copy:
    src: "/etc/morio/moriod/version*"
    dest: "{{ ansible_env.HOME }}/.moriod/"
    remote_src: yes
  ignore_errors: true
```

There may be a reason that you're doing it like this. And if so, I'd like to learn about it.
But if not, I think simpler is better.

File: ansible/playbooks/reset_moriod.yml [hunk 3]
```diff
@@ -11,34 +11,69 @@
...
       ignore_errors: true
 
-    - name: Remove Morio configuration files
+    - name: Find all files in the directory
+      ansible.builtin.find:
+        paths: /etc/morio/moriod
+        file_type: file
+      register: files_to_remove
+
+    - name: Remove all files in the directory
```
Comment (joostdecock) [inline ansible/playbooks/reset_moriod.yml:45]: Same comment here. This is split in a find + copy step, when it can be done in 1 step

File: ansible/playbooks/reset_moriod.yml [hunk 4]
```diff
@@ -11,34 +11,69 @@
...
       ignore_errors: true
 
-    - name: Remove Morio logs
+    - name: Find all log files in the directory
+      ansible.builtin.find:
+        paths: /var/log/morio/moriod
+        file_type: file
+      register: log_files
+
+    - name: Remove all log files
```
Comment (joostdecock) [inline ansible/playbooks/reset_moriod.yml:58]: Same comment here. This is split in a find + copy step, when it can be done in 1 step

File: ansible/playbooks/reset_moriod.yml (hunk 1 above)
Comment (joostdecock) [reply ansible/playbooks/reset_moriod.yml]: Same comment here. This is split in a find + copy step, when it can be done in 1 step

File: ansible/playbooks/reset_moriod.yml [hunk 5]
```diff
@@ -11,34 +11,69 @@
...
       ignore_errors: true
 
-    - name: Copy Back Version Files
+    - name: Find version files in backup
+      ansible.builtin.find:
+        paths: "{{ ansible_env.HOME }}/.moriod"
+        patterns: version*
+      register: backup_version_files
+
+    - name: Copy back version files
```
Comment (joostdecock) [inline ansible/playbooks/reset_moriod.yml:84]: Same comment here. This is split in a find + copy step, when it can be done in 1 step

File: package.json [hunk 6]
```diff
@@ -39,6 +39,7 @@
     "node": ">=20"
   },
   "scripts": {
+    "ephemeral:testvm": "node ./tests/10_ephemeral.test.mjs",
```
Comment (joostdecock) [inline package.json]: I think it makes more sense to give these run scripts a common prefix, rather than a common suffix.

I also don't want to type too much 😉 so how about we prefix things with `i` for integration tests and do:

- `ideploy` : Deploys test infrastructure
- `isetup` : Sets up morio on test infrastructure
- `itest` : Runs integration tests
- `ireset` : Resets morio on test infrastructure
- `idestroy` : Destroys/Removes test infrastructure

To run the ephemeral tests, you would just specify them with an argument:
```
npm run itest ephemeral
```

I think that would make it easier to remember all this.

Comment (joostdecock) [reply package.json]: Also, I think the things the run ansible should probably just call ansible directly, rather than going through NodeJS.

See my remarks below.

File: scripts/reset-test-vm.mjs [hunk 7]
```diff
@@ -25,15 +25,15 @@ const runAnsiblePlaybooks = async () => {
     }
 
     for (const dns of dnsNames) {
-      runPlaybook('stop_morio.yml', ansibleDir, dns);
+      runPlaybook('stop_moriod.yml', ansibleDir, dns);
```
Comment (joostdecock) [inline scripts/reset-test-vm.mjs]: This goes a bit against the way things are done in Ansible.
I'm sure it works, but it is going to be super confusing for people who are used to Ansible.

I think what would be a better way to go about this is to have our very first step (where we deploy the test infratstructure create an ansible inventory file.

Then, we can just run the various ansible playbooks and specify the targets. Ansible will handle all of the parallel processing for us.

Here is an example Ansible inventory file. I am using JSON here because that way we can also load this easily in NodeJS

```json
{
  "local":{
    "hosts":{
      "localhost":{
        
      }
    }
  },
  "all":{
    "children":{
      "pr98":{
        "hosts":{
          "abc.test.morio.it":{
            "example_variable":"example_value"
          }
        },
        "pr99":{
          "hosts":{
            "def.test.morio.it":{
              "example_variable":"example_value"
            },
            "ghi.test.morio.it":{
              "example_variable":"example_value"
            },
            "jkl.test.morio.it":{
              "example_variable":"example_value"
            }
          }
        }
      }
    }
  }
}
```

This allows us to combine ansible's `-i` (or `--inventory-file`) flag and its `-l` flag (or `--limit`) to limit the playbook run to those hosts we want. Eg:

```
ansible-playbook -i ./ansible/inventory.json -l pr99 ./ansible/some-playbook.yml
```

Or for improved legibility:
```
ansible-playbook --inventory-file ./ansible/inventory.json --limit pr99 ./ansible/some-playbook.yml
```

Comment (joostdecock) [reply scripts/reset-test-vm.mjs]: Also, as I am writing this, a thought occurred to me: Perhaps using Terraform to deploy the initial infrastructure is overkill 🤔 

Terraform really shines when keeping infrastructure in sync with your desired configuration over time. But for that, we would need to store the terraform state file somewhere, something which we are currently not doing.

However, we do not really need consistency over time. We can just create and destroy, using labels/tags to ensure everything gets cleaned up.

So perhaps it makes more sense to deploy the infrastructure with ansible, using the aws module: https://docs.ansible.com/ansible/latest/collections/amazon/aws/index.html

This will mean one less thing (terraform) to manage and have to understand before one can contributor meaningfully to the project.

Curious to hear your thoughts on this.

File: scripts/setup-test-vm.mjs [hunk 8]
```diff
@@ -42,7 +42,7 @@ const runAnsiblePlaybook = async () => {
     // Run Ansible playbook for each node
     for (const dns of dnsNames) {
       console.log(`Running Ansible playbook for node: ${dns}`);
-      const ansibleCommand = `ansible-playbook -i ${dns}, --extra-vars "ansible_ssh_private_key_file=~/.ssh/id_rsa ansible_user=admin ansible_ssh_common_args='-o StrictHostKeyChecking=no'" ../ansible/playbooks/install_morio.yml`;
+      const ansibleCommand = `ansible-playbook -i ${dns}, --extra-vars "ansible_ssh_private_key_file=~/.ssh/id_rsa ansible_user=admin ansible_ssh_common_args='-o StrictHostKeyChecking=no'" ../ansible/playbooks/install_moriod.yml`;
```
Comment (joostdecock) [inline scripts/setup-test-vm.mjs]: I think that as I explained above, this makes more sense to run directly, rather than through NodeJS

File: tests/10_ephemeral.test.mjs [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+import { pkg, corePkg } from '../api/tests/json-loader.mjs'
+import { errors } from '../api/src/errors.mjs'
+import dotenv from 'dotenv';
+
+dotenv.config();
+
+const api = restClient(`https://${process.env.MORIO_TEST_HOST}/-/api`)
+
+describe('Ephemeral API: Status Routes', () => {
+  /*
```
Comment (joostdecock) [inline tests/10_ephemeral.test.mjs]: I have included this example response in my original tests, but I think it's de-facto impossible to maintain.
So I would suggest removing it.

File: tests/utils.mjs [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+  try {
+    const response = await axios.get(url, requestConfig);
+    await pipeline(response.data, res);
+  } catch (err) {
+    console.error(err);
+  }
+}
+
+async function __postput(method, url, data, raw = false, log = false, ignoreCertificate = true) {
```
Comment (joostdecock) [inline tests/utils.mjs]: In a more recent project, I have taken to naming this `__withBody`, and `__withoutBody`.
Because `__postput` is not a good name, as `patch` is also with body, and `delete` is without body.

File: tests/utils.mjs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+    const body = raw ? response.data : response.data;
+    return [response.status, body];
+  } catch (err) {
+    if (log) console.error({ url, err });
+    return [err.status, err.response.data];
+  }
+}
+
+export async function streamGet(url, res, ignoreCertificate = true) {
```
Comment (joostdecock) [inline tests/utils.mjs]: This can go. I do not think we have streaming responses in the API.

File: tests/utils.mjs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        return ![4, 5].includes(String(result.status).charAt(0));
+      default:
+        return result;
+    }
+  } catch (err) {
+    return options.returnError ? err : false;
+  }
+}
+
+export async function get(url, raw = false, log = false, ignoreCertificate = true) {
```
Comment (joostdecock) [inline tests/utils.mjs]: I would call this `__withoutBody` (see below)

**Review by joostdecock — CHANGES_REQUESTED** [review]

File: ansible/playbooks/reset_moriod.yml (hunk 1 above)
Comment (levente232) [reply ansible/playbooks/reset_moriod.yml]: I rewrote "Remove all database files" to "Remove all data files".

split remove step into 2 "find + remove" because ansible builtin file module do not support direct use of pattern like "version*".

File: ansible/playbooks/reset_moriod.yml (hunk 2 above)
Comment (levente232) [reply ansible/playbooks/reset_moriod.yml:31]: split remove step into 2 "find + remove" because ansible builtin file module do not support direct use of pattern like "version*".

File: ansible/playbooks/reset_moriod.yml (hunk 4 above)
Comment (levente232) [reply ansible/playbooks/reset_moriod.yml:58]: split copy step into 2 "find + copy" because ansible builtin file module do not support direct use of pattern like "version*".

File: ansible/playbooks/reset_moriod.yml (hunk 3 above)
Comment (levente232) [reply ansible/playbooks/reset_moriod.yml:45]: split remove step into 2 "find + remove" because ansible builtin file module do not support direct use of pattern like "version*".

File: ansible/playbooks/reset_moriod.yml (hunk 5 above)
Comment (levente232) [reply ansible/playbooks/reset_moriod.yml:84]: split copy step into 2 "find + copy" because ansible builtin file module do not support direct use of pattern like "version*".

File: tests/utils.mjs (hunk 11 above)
Comment (levente232) [reply tests/utils.mjs]: Yes, I have removed it.

File: tests/utils.mjs (hunk 12 above)
Comment (levente232) [reply tests/utils.mjs]: I've changed the get function name to __withoutBody and give first parameter method("GET" or "DELETE") and modified function logic fit for both get and delete.

File: tests/utils.mjs (hunk 10 above)
Comment (levente232) [reply tests/utils.mjs]: I have changed the name from __postput to __withBody.

File: tests/10_ephemeral.test.mjs (hunk 9 above)
Comment (levente232) [reply tests/10_ephemeral.test.mjs]: I've removed it.

File: scripts/setup-test-vm.mjs (hunk 8 above)
Comment (levente232) [reply scripts/setup-test-vm.mjs]: "isetup": "dotenv -e .env -- sh -c 'ansible-playbook -i \"$MORIO_TEST_DNS_NAMES\" --extra-vars \"ansible_ssh_private_key_file=~/.ssh/id_rsa ansible_user=admin ansible_ssh_common_args=\\\"-o StrictHostKeyChecking=no\\\"\" ./ansible/playbooks/install_moriod.yml'",

I added this script to package.json file so I can run ansible-playbook directly.

File: package.json (hunk 6 above)
Comment (levente232) [reply package.json]: Modify all the scripts trigger names and usage.
- npm run ideploy
- npm run isetup
- npm run iconfigure
- npm run itest ephemeral(test name)
- npm run idestroy

- npm run ireset

File: scripts/reset-test-vm.mjs (hunk 7 above)
Comment (levente232) [reply scripts/reset-test-vm.mjs]: I've changed all deploy, setup, destroy, reset scenario as a direct ansible playbook.

File: .github/workflows/test-deploy.yml [hunk 13]
```diff
@@ -0,0 +1,73 @@
+name: Deploy and Configure Morio
```
Comment (joostdecock) [inline .github/workflows/test-deploy.yml]: Can we name this file `itest-deploy.yml`, just to differentiate between these integration tests (itest) and other tests.

File: .github/workflows/test-deploy.yml [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version: '20'  # Replace with the Node.js version you need
+
+      - name: Install AWS CLI
```
Comment (joostdecock) [inline .github/workflows/test-deploy.yml]: GitHub Actions runners come with the aws cli pre-installed 😇  no need to install it

File: .github/workflows/test-deploy.yml [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+          aws --version
+
+      - name: Configure AWS CLI
+        run: |
+          aws configure set aws_access_key_id ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws configure set aws_secret_access_key ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws configure set region ${{ secrets.AWS_REGION }}  # You can specify the region here, e.g., 'us-west-2'
+          aws configure set output json  # Set the output format to JSON
+
+      - name: Install Python 3 and pip
```
Comment (joostdecock) [inline .github/workflows/test-deploy.yml]: GitHub Actions runners also come with Pything pre-installed 😇

File: ansible/playbooks/deploy_moriod.yml [hunk 16]
```diff
@@ -0,0 +1,118 @@
+---
+- name: Provision EC2 Instances and Configure DNS
+  hosts: localhost
+  gather_facts: no
+  collections:
+    - amazon.aws
+    - community.general
+  vars:
+    region: "eu-central-1"
```
Comment (joostdecock) [inline ansible/playbooks/deploy_moriod.yml:10]: Please read all of these values from the Github Actions secrets:

- `subnet_id`: `${{ secrets.ITEST_AWS_SUBNET_ID }}`
- `security_group_id`: `${{ secrets.ITEST_AWS_SECURITY_GROUP_ID }}`
- `zone_id`: `${{ secrets.ITEST_AWS_ROUTE53_ZONE_ID }}`
- `ami_id`: `${{ secrets.ITEST_AWS_DEBIAN12_AMI_ID }}` (we will later also test on redhat-based distros)

In general, keep in mind that the threat model for CERT-EU is very different from your average organisation.
So in this case, we will rotate all of these so that the values in this PR become irrelevant.

Long story short: Do not put anything in the repo that can give attackers an advantage, no matter how small.

File: ansible/playbooks/deploy_moriod.yml [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        dns_names: []
+        
+    - name: Generate random DNS names and append to list
+      set_fact:
+        dns_names: "{{ dns_names + [lookup('community.general.random_string', count=1, length=3, special=False, upper=False)] }}"
+      loop: "{{ range(node_count)|list }}"
+
+    - name: Create an SSH Key Pair
+      amazon.aws.ec2_key:
+        name: "morio_test_{{ pr_id }}-{{ item }}"
```
Comment (joostdecock) [inline ansible/playbooks/deploy_moriod.yml]: Please use `morio_itest_....` here too, instead of `morio_test....`

File: ansible/playbooks/deploy_moriod.yml [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        key_material: "{{ lookup('file', '~/.ssh/id_rsa.pub') }}"
+      loop: "{{ dns_names }}"
+
+    - name: Create EC2 instances
+      amazon.aws.ec2_instance:
+        name: "morio-test-{{ pr_id }}-{{ item }}"
```
Comment (joostdecock) [inline ansible/playbooks/deploy_moriod.yml]: And here too (itest instead of test)

File: ansible/playbooks/deploy_moriod.yml [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+        key_name: "morio_test_{{ pr_id }}-{{ item }}"
```
Comment (joostdecock) [inline ansible/playbooks/deploy_moriod.yml]: Here too (itest instead of test)

File: ansible/playbooks/deploy_moriod.yml [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+        instance_type: "t3.medium"
+        image_id: "{{ ami_id }}"
+        region: "{{ region }}"
+        vpc_subnet_id: "{{ subnet_id }}"
+        security_group: "{{ security_group_id }}"
+        network:
+          assign_public_ip: true
+        wait: yes
+        tags:
+          Name: "morio-test-{{ pr_id }}-{{ item }}"
```
Comment (joostdecock) [inline ansible/playbooks/deploy_moriod.yml]: Here too (itest instead of test)

File: ansible/playbooks/deploy_moriod.yml [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+          morio: "true"
+          pull_request: "{{ pr_id }}"
+        user_data: |
+          #!/bin/bash
+          echo "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIEe3xiWGZ6u7h4/Z+MHmvk81uYNVSTlXIGR1bpyUEoa6 admin@cert.europa.eu" >> /home/admin/.ssh/authorized_keys
```
Comment (joostdecock) [inline ansible/playbooks/deploy_moriod.yml]: Let's put these SSH keys in secrets too ok

**Review by joostdecock — CHANGES_REQUESTED** [review]

File: .github/workflows/test-deploy.yml (hunk 13 above)
Comment (levente232) [reply .github/workflows/test-deploy.yml]: I've changed the name from test->itest

File: .github/workflows/test-deploy.yml (hunk 14 above)
Comment (levente232) [reply .github/workflows/test-deploy.yml]: I've removed it. :)

File: .github/workflows/test-deploy.yml (hunk 15 above)
Comment (levente232) [reply .github/workflows/test-deploy.yml]: I've removed it also. :)

File: ansible/playbooks/deploy_moriod.yml (hunk 16 above)
Comment (levente232) [reply ansible/playbooks/deploy_moriod.yml:10]: I've used Github Actions secrets instead.

File: ansible/playbooks/deploy_moriod.yml (hunk 18 above)
Comment (levente232) [reply ansible/playbooks/deploy_moriod.yml]: Corrected.

File: ansible/playbooks/deploy_moriod.yml (hunk 19 above)
Comment (levente232) [reply ansible/playbooks/deploy_moriod.yml]: Corrected.

File: ansible/playbooks/deploy_moriod.yml (hunk 17 above)
Comment (levente232) [reply ansible/playbooks/deploy_moriod.yml]: Corrected.

File: ansible/playbooks/deploy_moriod.yml (hunk 21 above)
Comment (levente232) [reply ansible/playbooks/deploy_moriod.yml]: ssh_pubkey_1: "{{ lookup('env', 'ITEST_AWS_SSH_PUBKEY_1_ID') }}"
    ssh_pubkey_2: "{{ lookup('env', 'ITEST_AWS_SSH_PUBKEY_2_ID') }}"
    
    I've used it from secrets too.

File: ansible/playbooks/deploy_moriod.yml (hunk 20 above)
Comment (levente232) [reply ansible/playbooks/deploy_moriod.yml]: Corrected.

**Review by joostdecock — APPROVED** [review]
